pulsar-client-cpp/lib/MessageId.cc

```diff
@@ -43,12 +52,64 @@ MessageId::MessageId(int64_t ledgerId, int64_t entryId)
      // consumer's partition is assigned to this partition
 }
 
+int64_t MessageId::getBatchIndex() const {
+    // It's only relevant for batch message ids
+    return -1;
+}
+
+const MessageId& MessageId::earliest() {
+    static const BatchMessageId _earliest(-1, -1);
+    return _earliest;
+}
+
+const MessageId& MessageId::latest() {
+    // For entry-id we only have 48bits
+    static const BatchMessageId _latest(std::numeric_limits<int64_t>::max(),
+                                        (int64_t)(pow(2, 47) - 1));
+    return _latest;
+}
+
+void MessageId::serialize(std::string& result) const {
+    proto::MessageIdData idData;
+    idData.set_ledgerid(ledgerId_);
+    idData.set_entryid(entryId_);
+    if (partition_ != -1) {
+        idData.set_partition(partition_);
+    }
+
+    idData.SerializeToString(&result);
+}
+
+/**
+ * Deserialize a message id from a binary string
+ */
+boost::shared_ptr<MessageId> MessageId::deserialize(const std::string& serializedMessageId) {
+    proto::MessageIdData idData;
+    if (!idData.ParseFromString(serializedMessageId)) {
+        throw "Failed to parse serialized message id";
+    }
+
+    return boost::make_shared<BatchMessageId>(idData.ledgerid(), idData.entryid(),
+                                              idData.batch_index());
+}
+
 
 #pragma GCC visibility push(default)
 std::ostream& operator<<(std::ostream& s, const pulsar::MessageId& messageId) {
     s << '(' << messageId.ledgerId_ << ',' << messageId.entryId_ << ',' << messageId.partition_ << ')';
     return s;
 }
+
+bool MessageId::operator<(const MessageId& other) const {
+    return std::tie(ledgerId_, entryId_, partition_)
+            < std::tie(other.ledgerId_, other.entryId_, other.partition_);
```

Comment:
MessageId.cc:110: error: cannot bind bitfield ‘other->pulsar::MessageId::entryId_’ to ‘int64_t&’

Reply:
Removed the `std::tie()` to support older g++

Reply:
Need to remove std::tie()